rllib/algorithms/dqn/torch/dqn_rainbow_torch_rl_module.py

```diff
@@ -0,0 +1,319 @@
+from ray.rllib.core.models.base import Encoder, ENCODER_OUT, Model
+from ray.rllib.core.rl_module.torch.torch_rl_module import TorchRLModule
+from ray.rllib.core.rl_module.rl_module import RLModule
+from ray.rllib.core.rl_module.rl_module_with_target_networks_interface import (
+    RLModuleWithTargetNetworksInterface,
+)
+from ray.rllib.policy.sample_batch import SampleBatch
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.framework import try_import_torch
+from ray.rllib.utils.typing import NetworkType, TensorType, TensorStructType
+
+torch, nn = try_import_torch()
+
+
+class DQNRainbowTorchRLModule(TorchRLModule, DQNRainbowRLModule):
+    framework: str = "torch"
+
+    @override(RLModule)
+    def _forward_inference(self, batch: Dict[str, TensorType]) -> Dict[str, TensorType]:
+        output = {}
+
+        # Encoder forward pass.
+        encoder_outs = self.encoder(batch)
+
+        # Q-head.
+        qf_outs = self._qf(encoder_outs[ENCODER_OUT])
+
+        # Get action distribution.
+        action_dist_cls = self.get_exploration_action_dist_cls()
+        action_dist = action_dist_cls.from_logits(qf_outs[QF_PREDS])
+        # Note, the deterministic version of the categorical distribution
+        # outputs directly the `argmax` of the logits.
+        exploit_actions = action_dist.to_deterministic().sample()
+
+        # In inference we only need the exploitation actions.
+        output[SampleBatch.ACTIONS] = exploit_actions
+
+        return output
+
+    # TODO (simon): Implement the timestep for more complex exploration
+    # schedules with epsilon greedy.
+    @override(RLModule)
+    def _forward_exploration(
+        self, batch: Dict[str, TensorType]
+    ) -> Dict[str, TensorType]:
+        output = {}
+
+        # Q-network forward pass.
+        qf_outs = self._qf(batch)
+
+        # Get action distribution.
+        action_dist_cls = self.get_exploration_action_dist_cls()
+        action_dist = action_dist_cls.from_logits(qf_outs[QF_PREDS])
+        # Note, the deterministic version of the categorical distribution
+        # outputs directly the `argmax` of the logits.
+        exploit_actions = action_dist.to_deterministic().sample()
+
+        # Apply epsilon-greedy exploration.
+        # TODO (simon): Implement sampling for nested spaces.
+        # TODO (simon): Implement different epsilon and schedules.
+        epsilon = 0.1
+        B = qf_outs[QF_PREDS].shape[0]
+        random_actions = torch.squeeze(
+            torch.multinomial(
+                (torch.nan_to_num(qf_outs[QF_PREDS], neginf=0.0) != 0).float(),
+                num_samples=1,
+            ),
+            dim=1,
+        )
+        output[SampleBatch.ACTIONS] = torch.where(
+            torch.rand((B,)) < epsilon,
+            random_actions,
+            exploit_actions,
+        )
+
+        return output
+
+    # TODO (simon): Maybe returning results in a dict of dicts:
```

Comment:
@sven1977 I am interested in you opinion on using here rather a nested dictionary (not necessarily a `NestedDict` ;) ) instead of the plain one.

Reply:
Plain one. I'm trying hard right now to get rid of both NestedDict and SampleBatch/MultiAgentBatch. Wherever normal dicts already work, let's use them to simplify the code base.

Reply:
I think what's left to do:
* Learner.update() only accepts a MultiAgentBatch
* RLModules forward passes only accept `NestedDict`, but I'm not 100% sure. Normal dicts might also work already here. We just have to convert to tensors first, of course, but that's something that NestedDict also does not do automatically.

Reply:
So far normal dicts work. Tried this out for PPO, SAC and DQN Rainbow. 

In regard to the `output_specs` being plain: the `ActorCriticEncoder` works with a nested one (e.g. `output[ENCODER][CRITIC]`).